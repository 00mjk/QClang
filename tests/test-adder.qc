/* gates transcribed from qelib1.inc */
qubit tdg(qubit a)
{
	float pi;
	
	pi = 3.14159265359;
	return U(0., 0., -pi / 4., a);
}

qubit t(qubit a)
{
	float pi;

	pi = 3.14159265359;
	return U(0., 0., pi / 4., a);
}

tup(qubit, qubit, qubit) CCX(qubit a, qubit b, qubit c)
{
	(b, c) = CX(b, hadamard(c)); c = tdg(c);
	(a, c) = CX(a, c); c = t(c);
	(b, c) = CX(b, c); c = tdg(c);
	(a, c) = CX(a, c); b = t(b); c = hadamard(t(c));
	(a, b) = CX(a, b);
	(a, b) = CX(t(a), tdg(b));
	return (a, b, c);
}

/* adapted from QISKit/examples/generic/adder.qasm */

tup(qubit, qubit, qubit) majority(qubit a, qubit b, qubit c)
{
	(c, b) = CX(c, b);
	(c, a) = CX(c, a);
	return CCX(a, b, c);
}

tup(qubit, qubit, qubit) unmaj(qubit a, qubit b, qubit c)
{
	(a, b, c) = CCX(a, b, c);
	(c, a) = CX(c, a);
	(a, b) = CX(a, b);
	return (a, b, c);
}

tup(qubit[], qubit) ripple_add(qubit[] a, qubit[] b)
{
	qubit carry;
	qubit cout;
	qubit[] out;
	int i;

	out = new qubit[](length(a));
	carry = false;
	for (i = 0; i < length(a); i = i + 1) {
		(b[i], a[i], carry) = majority(carry, b[i], a[i]);
	}

	(carry, cout) = CX(carry, false);

	for (i = length(a) - 1; i >= 0; i = i - 1) {
		(carry, out[i], a[i]) = unmaj(b[i], a[i], carry);
	}

	return (out, cout);
}

bit[] meas_array(qubit[] a) {
	int i;
	bit[] out;

	out = new bit[](length(a));
	for (i = 0; i < length(a); i = i + 1) {
		out[i] = measure(a[i]);
	}

	return out;
}

int main() {
	qubit[] a;
	qubit[] b;
	qubit cout;

	a = new qubit[](4);
	b = new qubit[](4);

	a[3] = false; a[2] = false; a[1] = false; a[0] = true; /* a = 0001 */
	b[3] = true; b[2] = true; b[1] = true; b[0] = true; /* b = 1111 */
	(b, cout) = ripple_add(a, b);
	meas_array(b);
	measure(cout);
	return 0;
}
