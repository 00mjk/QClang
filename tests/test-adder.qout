COMPLETED
{'1 0 0 0 0 0': 1}
dict_keys(['1 0 0 0 0 0'])
