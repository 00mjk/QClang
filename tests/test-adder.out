OPENQASM 2.0;
include "qelib1.inc";
qreg q[1];
creg c[1];
h q;
qreg temp_8[1];
qreg temp_9[1];
qreg temp_10[1];
qreg temp_11[1];
x temp_11;
qreg temp_12[1];
x temp_12;
qreg temp_13[1];
x temp_13;
qreg temp_14[1];
x temp_14;
qreg temp_15[1];
x temp_15;
qreg temp_21[1];
cx temp_11, temp_15;
cx temp_11, temp_21;
h temp_11;
cx temp_15, temp_11;
U(0., 0., -0.785398163398) temp_11;
cx temp_21, temp_11;
U(0., 0., 0.785398163398) temp_11;
cx temp_15, temp_11;
U(0., 0., -0.785398163398) temp_11;
cx temp_21, temp_11;
U(0., 0., 0.785398163398) temp_15;
U(0., 0., 0.785398163398) temp_11;
h temp_11;
cx temp_21, temp_15;
U(0., 0., -0.785398163398) temp_15;
U(0., 0., 0.785398163398) temp_21;
cx temp_21, temp_15;
cx temp_10, temp_14;
cx temp_10, temp_11;
h temp_10;
cx temp_14, temp_10;
U(0., 0., -0.785398163398) temp_10;
cx temp_11, temp_10;
U(0., 0., 0.785398163398) temp_10;
cx temp_14, temp_10;
U(0., 0., -0.785398163398) temp_10;
cx temp_11, temp_10;
U(0., 0., 0.785398163398) temp_14;
U(0., 0., 0.785398163398) temp_10;
h temp_10;
cx temp_11, temp_14;
U(0., 0., -0.785398163398) temp_14;
U(0., 0., 0.785398163398) temp_11;
cx temp_11, temp_14;
cx temp_9, temp_13;
cx temp_9, temp_10;
h temp_9;
cx temp_13, temp_9;
U(0., 0., -0.785398163398) temp_9;
cx temp_10, temp_9;
U(0., 0., 0.785398163398) temp_9;
cx temp_13, temp_9;
U(0., 0., -0.785398163398) temp_9;
cx temp_10, temp_9;
U(0., 0., 0.785398163398) temp_13;
U(0., 0., 0.785398163398) temp_9;
h temp_9;
cx temp_10, temp_13;
U(0., 0., -0.785398163398) temp_13;
U(0., 0., 0.785398163398) temp_10;
cx temp_10, temp_13;
cx temp_8, temp_12;
cx temp_8, temp_9;
h temp_8;
cx temp_12, temp_8;
U(0., 0., -0.785398163398) temp_8;
cx temp_9, temp_8;
U(0., 0., 0.785398163398) temp_8;
cx temp_12, temp_8;
U(0., 0., -0.785398163398) temp_8;
cx temp_9, temp_8;
U(0., 0., 0.785398163398) temp_12;
U(0., 0., 0.785398163398) temp_8;
h temp_8;
cx temp_9, temp_12;
U(0., 0., -0.785398163398) temp_12;
U(0., 0., 0.785398163398) temp_9;
cx temp_9, temp_12;
qreg temp_58[1];
cx temp_8, temp_58;
h temp_8;
cx temp_12, temp_8;
U(0., 0., -0.785398163398) temp_8;
cx temp_9, temp_8;
U(0., 0., 0.785398163398) temp_8;
cx temp_12, temp_8;
U(0., 0., -0.785398163398) temp_8;
cx temp_9, temp_8;
U(0., 0., 0.785398163398) temp_12;
U(0., 0., 0.785398163398) temp_8;
h temp_8;
cx temp_9, temp_12;
U(0., 0., -0.785398163398) temp_12;
U(0., 0., 0.785398163398) temp_9;
cx temp_9, temp_12;
cx temp_8, temp_9;
cx temp_9, temp_12;
h temp_9;
cx temp_13, temp_9;
U(0., 0., -0.785398163398) temp_9;
cx temp_10, temp_9;
U(0., 0., 0.785398163398) temp_9;
cx temp_13, temp_9;
U(0., 0., -0.785398163398) temp_9;
cx temp_10, temp_9;
U(0., 0., 0.785398163398) temp_13;
U(0., 0., 0.785398163398) temp_9;
h temp_9;
cx temp_10, temp_13;
U(0., 0., -0.785398163398) temp_13;
U(0., 0., 0.785398163398) temp_10;
cx temp_10, temp_13;
cx temp_9, temp_10;
cx temp_10, temp_13;
h temp_10;
cx temp_14, temp_10;
U(0., 0., -0.785398163398) temp_10;
cx temp_11, temp_10;
U(0., 0., 0.785398163398) temp_10;
cx temp_14, temp_10;
U(0., 0., -0.785398163398) temp_10;
cx temp_11, temp_10;
U(0., 0., 0.785398163398) temp_14;
U(0., 0., 0.785398163398) temp_10;
h temp_10;
cx temp_11, temp_14;
U(0., 0., -0.785398163398) temp_14;
U(0., 0., 0.785398163398) temp_11;
cx temp_11, temp_14;
cx temp_10, temp_11;
cx temp_11, temp_14;
h temp_11;
cx temp_15, temp_11;
U(0., 0., -0.785398163398) temp_11;
cx temp_21, temp_11;
U(0., 0., 0.785398163398) temp_11;
cx temp_15, temp_11;
U(0., 0., -0.785398163398) temp_11;
cx temp_21, temp_11;
U(0., 0., 0.785398163398) temp_15;
U(0., 0., 0.785398163398) temp_11;
h temp_11;
cx temp_21, temp_15;
U(0., 0., -0.785398163398) temp_15;
U(0., 0., 0.785398163398) temp_21;
cx temp_21, temp_15;
cx temp_11, temp_21;
cx temp_21, temp_15;
creg temp_15_mb100[1];
measure temp_15 -> temp_15_mb100;
creg temp_14_mb101[1];
measure temp_14 -> temp_14_mb101;
creg temp_13_mb102[1];
measure temp_13 -> temp_13_mb102;
creg temp_12_mb103[1];
measure temp_12 -> temp_12_mb103;
creg temp_58_mb104[1];
measure temp_58 -> temp_58_mb104;
